pythonforandroid/bootstrap.py

```diff
@@ -257,6 +266,83 @@ def _unpack_aar(self, aar, arch):
+
+        libs_to_load = []
+        # Add libs that should be loaded before bootstrap libs (here goes some
+        # special shared libs, like stl libs used at compile time or libs that
+        # the bootstrap may be dependant).
+        # Todo: All the recipes using gnustl_shared should be migrated to use
+        # c++_shared because in ndk r18 it will be the only stl available:
+        # https://developer.android.com/ndk/guides/cpp-support
+        for recipe_name in self.ctx.recipe_build_order:
+            recipe = Recipe.get_recipe(recipe_name, self.ctx)
+            if recipe.stl_depends is None:
+                continue
+            if recipe.stl_depends not in ('c++_shared', 'gnustl_shared'):
+                raise BuildInterruptingException(
+                    'You have a non supported value in attribute stl_depends '
+                    '"{stl_lib}" for recipe "{recipe_name}". The'
+                    'value should be one of: "c++_shared" or "gnustl_shared"'.format(
+                        stl_lib=recipe.stl_depends, recipe_name=recipe.name))
+            if recipe.stl_depends not in libs_to_load:
+                libs_to_load.append(recipe.stl_depends)
+            if all(lib in libs_to_load for lib in ('c++_shared', 'gnustl_shared')):
+                break
+
+        # Add crystax if necessary
+        if self.ctx.python_recipe.from_crystax:
+            libs_to_load.append('crystax')
+
+        # Add libraries that python depends on
+        if 'sqlite3' in self.ctx.recipe_build_order:
+            libs_to_load.append('sqlite3')
+        if 'libffi' in self.ctx.recipe_build_order:
+            libs_to_load.append('ffi')
+        if 'openssl' in self.ctx.recipe_build_order:
+            recipe = Recipe.get_recipe('openssl', self.ctx)
+            libs_to_load.append('ssl' + recipe.version)
+            libs_to_load.append('crypto' + recipe.version)
+
+        # Add the bootstrap libs
+        [libs_to_load.append(i) for i in self.libraries_to_load]
+
+        # Add the corresponding python lib
+        python_version = self.ctx.python_recipe.major_minor_version_string
+        if python_version[0] == '3':
+            python_version += 'm'
+        libs_to_load.append('python' + python_version)
+
+        # Add libs that should be loaded after python because depends on it
+        if 'boost' in self.ctx.recipe_build_order:
+            # Todo: check boost loading dependencies and create a proper
+            # loading order. For now it works for libtorrent's recipe, the only
+            # recipe that depends on boost...so...for now it's fine, but should
+            # be enhanced, maybe using similar method than
+            # `pythonforandroid.build.copylibs_function`?
+            for l in listdir(self.ctx.get_libs_dir(arch.arch)):
+                if l.startswith('libboost_'):
+                    libs_to_load.append(l[3:-3])
+        if 'libtorrent' in self.ctx.recipe_build_order:
```

Comment:
I don't think it makes much sense to have libtorrent/libiconv/libzbar code hardcoded here. `boost`, *maybe*, but even that is a relatively unimportant lib for the p4a ecosystem as a whole. Would it be possible to somehow also put this into the respective recipes and have it accessed here?

But in general, I also don't quite understand why preloading libs that should be linked dynamically (and hence automatically loaded by the OS loader of Android) is necessary in the first place. Do you know why this is required? I'm very curious

Reply:
> Do you know why this is required?

Ok, this is what I think is happening and is based on my experience, maybe android is capable to load the system libs but I suspect that android os has no way to know the library order loading, for instance, take the libpython3.X.so as example and imagine that we build with our libffi.so. Given that situation if  android loads first the libpython3X.so it will crash because we build it as a shared library and at the time that the library gets loaded will not have all the functions/symbols needed declared in libffi.so, because still is not loaded, so the application will crash. I also use the libtorrent since introduced in p4a and that also applies to it...boost must be loaded after python (because depends on python) and libtorrent after boost libraries (because depends on boost)...if the order is not right...buuuum...crash complaining about missing symbols/functions. Those are the cases I remember..there is more. So...I suspect that the android os has no way to know the right loading order for external libraries.

Reply:
That sounds to me like the dynamic link is just wrong. E.g. you built libtorrent wrong without properly depending on libboost - like the main program depends on it properly, but your `libtorrent.so` doesn't. I highly doubt the OS loader in Android cannot resolve this properly, since it's just the Linux loader, so IMHO this should be investigated a lot deeper before we merge this

Reply:
Also a common problem: maybe it depends on libboost, but the wrong version or with the wrong library file name referenced? It could be any of those things, and we should be very sure that's not the culprit before we add all this extra complexity

Reply:
@JonasT, I don't think that I have linkage problems with the boost/libtorrent recipe, I already tested in one of my apps (which integrates a torrent client, running with libtorrent rasterbar) and I downloaded some file without issues...well...only one problem...the version I introduce is a release candidate version and some changes has been made in the source code that may require to adapt some calls inside the app code...but this has nothing to do with this talk. Also I should mention that the libtorrrent recipe is kind of special:

  - It is build with the standalone toolchain (which isolates an android environment with the proper files...so it's easier to set the proper flags because all the environment files are in one place)
  - also contains two libraries: One is the library which contains all the functions/symbols like any standard library and **It's not python dependant**. The other one has the python bindings and that one is installed into site-packages like a regular python module (and that one does not require any special treatment). The first one should be treated as a regular library, that is why we need to load it like any standard library.

Saying that, if you wanna take a look into the libtorrent recipe I created a pr #1555, You will see the relevant code in there and despite it works, I may did something wrong with the flags or with something else...so you tell me please...:smile:...I will fix it if that is the case.

Reply:
Well again, if it doesn't load at runtime then your dynamic link is most likely wrong.

The recipe isn't that interesting, what would be interesting if you used `nm` and `objdump` to actually check on your produced `.so` files to see what they link, and if these links are correct. (Most notably, do they link the libraries you hacked into your `libraries_to_load` mechanism, with exactly the correct name? I doubt they do, otherwise it shouldn't be necessary)

Reply:
Can you give me an affected `.apk` maybe? One that doesn't use `libraries_to_load` and therefore doesn't work? You could e-mail it to `jonas@thiem.email` and I could try `nm` and `objdump` to see if i can find out what exactly is wrong with the linking, or why it doesn't launch properly

Reply:
@JonasT, no need that pass you any application, I just made some test it in a new branch (using the branch python-core-update **on his current status** (pr #1537 as a base),  adding the update of the libtorrent recipe and adapting one of the test apps (the one with openssl and sqlite3 for python3) to libtorrent instead of using the numpy module and...whatever was happening before...it's not happening now :smile: ...so...good news for all of us!!!

I suspect that my last commit in the python-core-update branch (bdd2b7c) maybe related with this, because now it works the ctypes module without importing libffi...and I did the pr for the libffi's python3 compatibility (#1463) and at that time it was failing...not now using the new core...

In my last test I comment all the libs, only with the libpythonX.X.so and libmain.so seems that the test app works...

If we remove the libpythonX.X.so (with only the libmain.so in PythonUtil.java) then the app not starts and I see some error in logcat similar to:

`dlopen failed: could not load library "libpython3.7m.so.1.0"...`

*Note: We already seen something similar in the issue #1501...mmm*

Anyway, here are the branch I used to make the test:

  - https://github.com/opacam/python-for-android/commits/python-core-libs-loading-test

  - and the commit that I thing that fixed the libraries loading problems is bdd2b7c.

Still not sure about all this, let me do some more test with a real app but I suspect that we can remove most of the libs of the PythonUtil.java and probably drop this pr and maybe you can test some of your app without the ctypes fix (#1517) using the python-core-update branch (maybe we doesn't need that fix now). Also it would be great that you confirm my words in this post, so we can remove this pr.

Reply:
`dlopen failed: could not load library "libpython3.7m.so.1.0"` yes the problem here seems to be that this file exists in the distribution (the `.so`) but gradle somehow throws it out (gone from the `.apk`). So it should most likely be only `libmain.so`, and whatever goes wrong with gradle should be fixed / #1501 should be fixed

Reply:
@JonasT, seems to me that we do not need this pr
¿Do you want to create a new pr, something like `simplify loading mechanism` to remove the unneeded stuff? If you don't wanna to do it I could reconvert this pr to do so or create a new one...I ask you because maybe you already have some work done on this, you have been one step ahead all the way with this matter...so...

Reply:
@opacam no I have been just waiting... haven't touched this part. So feel free to update it, I really hope the others will wake up soon and actually do some merging